Make J&K or S&D options generic (can use any letter combination)

BlueDrink9 opened the pull request:
Adds a list option to dynamically create hotkey pairs like j & k,
allowing users to set an arbitrary combination (like kv, fd, etc).

Doesn't touch jj. Does remove SD and JK settings, which may need a deprecation warning if you care about changing settings.

Also includes some refactoring. You may want me to revert 
a5b2323, but I think it's less likely to lead to errors due to typos (at the expense of being slightly harder to read, but I think that could also still be fixed if you wanted it)

BlueDrink9 commented:
@rcmdnk do you know when you might have time to look at this?

rcmdnk commented:
sorry, I could not take enough time to check codes.

Now I tried to merge the pull request.
Could you please check https://github.com/rcmdnk/vim_ahk/tree/dev

- I want to remain `sd`/`jk` options for now for the backword compatibility.
(should be better if they are included in two-letter options, but there is no time to do now)

- https://github.com/rcmdnk/vim_ahk/pull/42/commits/6434687db2ae7c117877deab781fce141819ce84#diff-c889379841ab55b6720371aa6753136eR176

It seems `#If` is not needed, as long as I tested.
Could you please explain when it is needed or what happens if it is not defined.

- https://github.com/rcmdnk/vim_ahk/pull/42/commits/6434687db2ae7c117877deab781fce141819ce84#diff-c889379841ab55b6720371aa6753136eR20
Is this out of class function necessary or did you just forget to delete?

BlueDrink9 commented:
Would you like me to rebase it onto your dev branch to make it easier to merge?

BlueDrink9 commented:
Regarding the current sd and jk options, would you accept:

Reading the old options, and if they are set, add the relevant letters into the new box? That way anyone that has set them won't be affected, and new users won't see redundant options

rcmdnk commented:
https://github.com/rcmdnk/vim_ahk/tree/dev branch bases on your PR.
Therefore if you do not see the problem in there (see above comments),
I can merge directly (which will mark this PR as merged, I think),
or you can update this PR and I will merge.

About `jk`/`sd` options, 

>  if they are set, add the relevant letters into the new box?

this is great.
If you can work on it, I prefer to make another PR after merging this PR.

BlueDrink9 commented:
The out of class function was an accident I think. Deleted now, same as your dev branch.

I also added a commit to the .gitignore. You can add that manually.

[I think the change to use `ObjBindMethod` reduces readability a lot.](https://github.com/BlueDrink9/vim_ahk/compare/twoLetterNormalNew...rcmdnk:dev#diff-c889379841ab55b6720371aa6753136eR169)

I will merge your dev branch and update it with your requirements. I will removed all the `sd`/`jk` documentation, checkboxes etc so that new people won't see them, but update the code so that if you have them set in the `.ini`, your settings will be updated to use them via the new generic system. I will make a new PR for these changes, however.

So for now, this branch can be merged.

BlueDrink9 commented:
I wasn't sure what `#if` you were referring to, but both are definitely necessary. Unless your `objbindmethod` stuff is doing odd black magic, the documentation linked in the comment says why, and my branch will error if it is removed.

The plain `#if` was just to negate the previous `#if ...`

rcmdnk commented:
OK, I read the document.
If FunctionObject is used, `#if` is not needed.
There is no magic. 

ObjBindMethod should be used if class methods are defined.
Otherwise, we should revert to the previous version (no class, use labels instead of functions...)

rcmdnk commented:
About .gitignore,
I don't want to assume any kind of global .gitignore of others.
This is Windows' project, then .DS_Store may not seem proper, but I sometimes develop this in Mac, so .DS_Store is better to be there.

`!/lib` is harmless and can be there, but I think to put `/lib` in global gitignore is so aggressive because many of projects have pre-installed lib/ (especially for interpreter languages)

BlueDrink9 commented:
Ok, I understand and agree with what you are saying. I think this can be merged for now.

It's a side track, but I really think `.DS_store` is something you should assume global. Otherwise, every project would have to have it in case a user devs on Mac. On the other hand, if you develop on OSX, you need to ignore it for every project. That isn't the only file that has this property either. It really feels like poor practise to clutter a project's .gitignore like that. Again, this is a side track, not related to the PR

BlueDrink9 commented:
#43 is based off your dev branch, so can be merged into that before you merge dev into master.

rcmdnk commented:
this function was merged by #43

BlueDrink9 commented:
Thanks. I notice the changelog doesn't include it. Since it changes how users might interact with it and adds a feature, I think it should be mentioned

rcmdnk commented:
You are right. The README will be updated by #44.

I may first merge some of the commits in #44 then we can discuss remainings later.
